TestGrid documentation (#2602)
kyma-project/test-infra: docs/prow/prow-k8s-testgrid.md

# TestGrid

> [TestGrid](https://testgrid.k8s.io) is an interactive dashboard for viewing tests results in a grid. It parses JUnit reports for generating grid view from the tests.

Reviewer: ```suggestion
[TestGrid](https://testgrid.k8s.io) is an interactive dashboard for viewing tests results in a grid. It parses JUnit reports for generating a grid view from the tests.
```

There is one dashboard group called `kyma` which groups dashboards from both of the organizations: *kyma-project* and *kyma-incubator*. The dashboards start with the prefix name which refers to one of the organizations: *kyma* and *kyma-incubator*.

> TestGrid configuration is stored inside [kubernetes/test-infra](https://github.com/kubernetes/test-infra/tree/master/config/testgrids/kyma) repository in the `/config/testgrids/kyma/` directory.

Reviewer: ```suggestion
TestGrid configuration is stored inside the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/tree/master/config/testgrids/kyma) repository in the `/config/testgrids/kyma/` directory.
```

TestGrid is automatically updated by Prow using the tool called [transfigure.sh](https://github.com/kubernetes/test-infra/tree/master/testgrid/cmd/transfigure).

## Adding new dashboard

The dashboards' configuration is stored in the [testgrid-default.yaml](https://github.com/kyma-project/test-infra/tree/master/prow/testgrid-default.yaml) file.
This file is automatically generated from the [template testgrid-default.yaml](https://github.com/kyma-project/test-infra/blob/master/templates/templates/testgrid-default.yaml) file. See the example of a dashboard configuration file:
```yaml
dashboards:
  # kyma
  - name: kyma_integration
  - name: kyma_control-plane

  # kyma-incubator
  - name: kyma-incubator_compass

dashboard_groups:
  - name: kyma
    dashboard_names:
      - kyma_integration
      - kyma_control-plane
      - kyma-incubator_compass
```
Follow these steps to add a new dashboard:

> 1. Add a new dashboard definition in the `dashboards` field according to the example file above inside a template file. **Dashboards must have a dashboard group name as a prefix**.

Reviewer: ```suggestion
1. Add a new dashboard definition in the `dashboards` field inside a template file. **Dashboards must have a dashboard group name as a prefix**.
```

2. Add the previously added dashboard to the corresponding **dashboard_name** inside the **dashboard_groups** field.
3. Generate a new config file using the [rendertemplates](https://github.com/kyma-project/test-infra/tree/master/development/tools/cmd/rendertemplates) tool and check if the config file generated correctly.

## Adding ProwJob to the TestGrid

After adding a desired dashboard in the `testgrid-default.yaml` file, add ProwJobs to the dashboard. To do so, define a new **annotations** field in a ProwJob definition. See the example below:
```yaml
annotations:
  testgrid-dashboards: dashboard-name      # [Required] A dashboard already defined in a config.yaml.
  testgrid-tab-name: some-short-name       # [Optional] A shorter name for the tab. If omitted, just uses the job name.
  testgrid-alert-email: me@me.com          # [Optional] An alert email that will be applied to the tab created in the first dashboard specified in testgrid-dashboards.
  description: Words about your job.       # [Optional] A description of your job. If omitted, only the job name is used.

>   testgrid-num-columns-recent: "10"        # [Optional] The number of runs a row can be omitted from before it is considered stale. Currently defaults to 10.

Reviewer: ```suggestion
  testgrid-num-columns-recent: "10"        # [Optional] The number of runs in a row that can be omitted before the test is considered stale. The default value is 10.
```

  testgrid-num-failures-to-alert: "3"      # [Optional] The number of continuous failures before sending an email. The default value is 3.

>   testgrid-alert-stale-results-hours: "12" # [Optional] The number of hours that pass with no results after which the email is sent.

Reviewer: ```suggestion
  testgrid-alert-stale-results-hours: "12" # [Optional] The number of hours that pass with no results after which the email is sent. The default value is 12.
```

```

The only required field is **testgrid-dashboards**. It must correspond to the name defined in the `testgrid-default.yaml` file. It is also recommended to add a short description of the job in the **description** field.

> The rest of the fields is optional and can be omitted.

Reviewer: ```suggestion
The rest of the fields are optional and can be omitted.
```

TestGrid configuration is stored inside [kubernetes/test-infra](https://github.com/kubernetes/test-infra/tree/master/config/testgrids/kyma) repository in the `/config/testgrids/kyma/` directory.

TestGrid is being automatically updated by Prow using the tool called [transfigure.sh](https://github.com/kubernetes/test-infra/tree/master/testgrid/cmd/transfigure).

## Adding new dashboard

The dashboards' configuration is stored in the [testgrid-default.yaml](https://github.com/kyma-project/test-infra/tree/master/prow/testgrid-default.yaml) file.
This file is being automatically generated from a [template testgrid-default.yaml](https://github.com/kyma-project/test-infra/blob/master/templates/templates/testgrid-default.yaml) file.
```yaml
dashboards:
  # kyma
  - name: kyma_integration
  - name: kyma_control-plane

  # kyma-incubator
  - name: kyma-incubator_compass

dashboard_groups:
  - name: kyma
    dashboard_names:
      - kyma_integration
      - kyma_control-plane
      - kyma-incubator_compass
```
To add a new dashboard it's needed to do several steps:

> 1. Add a new dashboard definition in the `dashboards` field according to the example file above inside a template file. **Dashboards need to have dashboard name as prefix**.

Reviewer: ```suggestion
1. Add a new dashboard definition in the `dashboards` field inside a template file. **Dashboards must have a dashboard name as a prefix**.
```

Reviewer: Dashboard definition or configuration? 🤔

Author: It is a name of a dashboard that has to be used in the prowjob annotations. I may be *dashboard name* as well.

> 2. Add the previously added dashboard to the corresponding `dashboard_name` inside the `dashboard_groups` field.

Reviewer: ```suggestion
2. Add the previously added dashboard to the corresponding **dashboard_name** inside the **dashboard_groups** field.
```

> 3. Generate a new config file using [rendertemplates](https://github.com/kyma-project/test-infra/tree/master/development/tools/cmd/rendertemplates) tool and check if the config file generated correctly.

Reviewer: ```suggestion
3. Generate a new config file using the [rendertemplates](https://github.com/kyma-project/test-infra/tree/master/development/tools/cmd/rendertemplates) tool and check if the config file generated correctly.
```

## Adding ProwJob to the TestGrid

> After adding a desired dashboard in the `testgrid-default.yaml` file you need to add ProwJobs to the dashboard. It can be done by defining new `annotations` field to a ProwJob definition.

Reviewer: ```suggestion
After adding a desired dashboard in the `testgrid-default.yaml` file, add ProwJobs to the dashboard. To do so, define a new **annotations** field in a ProwJob definition.
```

Author: I think it would be best to leave the annotations word in a raw code brackets as it refers to the exact field which is placed in a prowjob definition.

> Add the below line to the ProwJob and edit it to your needs.

Reviewer: ```suggestion
See the example:
```

> The only required field is `testgrid-dashboards`. It needs to have a name that is defined in the `testgrid-default.yaml` file. It is also good to define `description` field to add short description about the job.

Reviewer: I'd move this paragraph below the example.

> The only required field is `testgrid-dashboards`. It needs to have a name that is defined in the `testgrid-default.yaml` file. It is also good to define `description` field to add short description about the job.

Reviewer: ```suggestion
The only required field is **testgrid-dashboards**. It must correspond to the name defined in the `testgrid-default.yaml` file. It is also recommended to add a short description of the job in the **description** field.
```

> The rest of the fields is optional and can be omitted.

Reviewer: ```suggestion
The rest of the fields are optional and can be omitted.
```

> ```yaml

Reviewer: I'd move this example to line 39.

annotations:

>   testgrid-dashboards: dashboard-name      # Required. a dashboard already defined in a config.yaml.

Reviewer: ```suggestion
  testgrid-dashboards: {DASHBOARD_NAME}      # [Required] This dashboard is already defined in the config.yaml.
```

>   testgrid-tab-name: some-short-name       # optionally, a shorter name for the tab. If omitted, just uses the job name.

Reviewer: ```suggestion
  testgrid-tab-name: {SHORT_NAME}       # [Optional] A shorter name for the tab. If omitted, only the job name is used.
```

>   testgrid-alert-email: me@me.com          # optionally, an alert email that will be applied to the tab created in the

Reviewer: ```suggestion
  testgrid-alert-email: {EMAIL_ADDRESS}          # [Optional] An alert email that will be applied to the tab created in the first dashboard specified in testgrid-dashboards.
```

>                                            # first dashboard specified in testgrid-dashboards.

Reviewer: ```suggestion

```

>   description: Words about your job.       # optionally, a description of your job. If omitted, just uses the job name.

Reviewer: ```suggestion
  description: {DESCRIPTION}       # [Optional] A description of your job. If omitted, only the job name is used.
```

>   testgrid-num-columns-recent: "10"        # optionally, the number of runs a row can be omitted from before it is

Reviewer: ```suggestion
  testgrid-num-columns-recent: "10"        # [Optional] The number of runs in a row that can be omitted before the test is considered stale. The default value is 10.
```

>                                            # considered stale. Currently defaults to 10.

Reviewer: ```suggestion

```

>   testgrid-num-failures-to-alert: "3"      # optionally, the number of continuous failures before sending an email.

Reviewer: ```suggestion
  testgrid-num-failures-to-alert: "3"      # [Optional] The number of continuous failures before sending an email. The default value is 3.
```

>                                            # Currently defaults to 3.

Reviewer: ```suggestion

```

>   testgrid-alert-stale-results-hours: "12" # optionally, send an email if this many hours pass with no results at all.

Reviewer: ```suggestion
  testgrid-alert-stale-results-hours: "12" # [Optional] The number of hours that pass with no results after which the email is sent.
```

```

> If the job does not have to be on a TestGrid use the annotation below to disable the generation of the TestGrid test group:

Reviewer: ```suggestion
If you don't want to include a job on the TestGrid, use this annotation to disable the generation of the TestGrid test group:
```

```yaml
annotations:
  testgrid-create-test-group: "false"
```

> This configuration generally applies for postsumbits and periodics. The presubmits jobs can be added to the TestGrid, however if there is no `annnotations` field defined the job will be omitted in config file generation.

Reviewer: ```suggestion
This configuration applies to postsumbit and periodic jobs. Presubmit jobs can be added to the TestGrid, however, if there is no **annnotations** field defined, the job will be omitted in the config file generation.
```
